File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        int rightMax = prices[n-1];
        int maxProfit = 0;
        
        for(int i = n-2; i >= 0; i--) {
            if(rightMax < prices[i]) rightMax = prices[i];
            else if(maxProfit < rightMax - prices[i]) maxProfit = rightMax - prices[i];
        }
        
        return maxProfit;
    }
};

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let minPrice = 1e8;
    let maxProfit = 0;
    
    for(let i = 0; i < prices.length; i++) {
        if(prices[i] < minPrice) 
            minPrice = prices[i];
        
        maxProfit = Math.max(maxProfit, prices[i] - minPrice);
    }
    
    return maxProfit;
};